cpp/src/arrow/dataset/file_base.cc

```diff
@@ -41,5 +44,56 @@ Status FileBasedDataFragment::Scan(std::shared_ptr<ScanContext> scan_context,
   return format_->ScanFile(source_, scan_options_, scan_context, out);
 }
 
+FileSystemBasedDataSource::FileSystemBasedDataSource(
+    fs::FileSystem* filesystem, const fs::Selector& selector,
+    std::shared_ptr<FileFormat> format, std::shared_ptr<ScanOptions> scan_options,
+    std::vector<fs::FileStats> stats)
+    : filesystem_(filesystem),
+      selector_(std::move(selector)),
+      format_(std::move(format)),
+      scan_options_(std::move(scan_options)),
+      stats_(std::move(stats)) {}
+
+Status FileSystemBasedDataSource::Make(fs::FileSystem* filesystem,
+                                       const fs::Selector& selector,
+                                       std::shared_ptr<FileFormat> format,
+                                       std::shared_ptr<ScanOptions> scan_options,
+                                       std::unique_ptr<FileSystemBasedDataSource>* out) {
+  std::vector<fs::FileStats> stats;
+  RETURN_NOT_OK(filesystem->GetTargetStats(selector, &stats));
+
+  auto new_end =
+      std::remove_if(stats.begin(), stats.end(), [&](const fs::FileStats& stats) {
+        return stats.type() != fs::FileType::File ||
+               !format->IsKnownExtension(stats.extension());
+      });
+  stats.resize(new_end - stats.begin());
+
+  out->reset(new FileSystemBasedDataSource(filesystem, selector, std::move(format),
+                                           std::move(scan_options), std::move(stats)));
+  return Status::OK();
+}
+
+std::unique_ptr<DataFragmentIterator> FileSystemBasedDataSource::GetFragments(
+    std::shared_ptr<ScanOptions> options) {
+  size_t i = 0;
+  auto stats = stats_;
+  auto scan_options = scan_options_;
+  auto format = format_;
+  auto filesystem = filesystem_;
+  return MakeIterator([=](std::shared_ptr<DataFragment>* out) mutable {
```

Comment:
The complexity of such succinct code is very high (4 captures, lifetime requirement, mutable!, output parameter...). I think it warrants creating an explicit Iterator class.

Reply:
Okay, can do.